From b0db060d0396ebc73d2e6533b1be6fe10236f711 Mon Sep 17 00:00:00 2001
From: Phil Sutter <psutter@redhat.com>
Date: Fri, 17 Jul 2026 11:12:20 +0200
Subject: [PATCH] evaluate: don't update cache for anonymous chains

JIRA: https://issues.redhat.com/browse/RHEL-190549
Upstream Status: nftables commit bd1fc6c740535d1ad3f38f8ba9c539c14f1732f3

commit bd1fc6c740535d1ad3f38f8ba9c539c14f1732f3
Author: Florian Westphal <fw@strlen.de>
Date:   Wed Mar 19 21:05:53 2025 +0100

    evaluate: don't update cache for anonymous chains

    Chain lookup needs a name, not a numerical id.
    After patch, loading bogon gives following errors:

    Error: No symbol type information a b index 1 10.1.26.a

    v2: Don't return an error, just make it a no-op (Pablo Neira Ayuso)

    Fixes: c330152b7f77 ("src: support for implicit chain bindings")
    Signed-off-by: Florian Westphal <fw@strlen.de>

Signed-off-by: Phil Sutter <psutter@redhat.com>
---
 src/evaluate.c                                            | 4 ++++
 .../bogons/nft-f/null_deref_on_anon_chain_update_crash    | 8 ++++++++
 2 files changed, 12 insertions(+)
 create mode 100644 tests/shell/testcases/bogons/nft-f/null_deref_on_anon_chain_update_crash

diff --git a/src/evaluate.c b/src/evaluate.c
index 710786f..5a7635d 100644
--- a/src/evaluate.c
+++ b/src/evaluate.c
@@ -5031,6 +5031,10 @@ static int rule_cache_update(struct eval_ctx *ctx, enum cmd_ops op)
 	if (!table)
 		return table_not_found(ctx);
 
+	/* chain is anonymous, adding new rules via index is not supported. */
+	if (!rule->handle.chain.name)
+		return 0;
+
 	chain = chain_cache_find(table, rule->handle.chain.name);
 	if (!chain)
 		return chain_not_found(ctx);
diff --git a/tests/shell/testcases/bogons/nft-f/null_deref_on_anon_chain_update_crash b/tests/shell/testcases/bogons/nft-f/null_deref_on_anon_chain_update_crash
new file mode 100644
index 0000000..310486c
--- /dev/null
+++ b/tests/shell/testcases/bogons/nft-f/null_deref_on_anon_chain_update_crash
@@ -0,0 +1,8 @@
+table ip f {
+        chain c {
+                jump {
+                        accept
+                }
+        }
+}
+a b index 1 10.1.26.a
